### ISSUE TITLE
Upgrading artifactory

### ISSUE THREAD

**Issue by michaeloa** [robertdebock/ansible-role-artifactory#3]
Ran into what I feel is an issue with the role in a playbook.

Upgraded the version in the playbook from 6.13.1 to 6.16.0 recently. Ran the playbook again.

This did not result in any upgrade of artifactory - even though I can see that it downloads the new files. I then noticed that the re-run of the playbook does not restart the service, so I tried to see what happens if you do that. The result was that the server restarted completely - wiping the contents of the old server and creating a new install.

That is not ideal behavior, obviously. I think that the role ought to either do the upgrade correctly - alternatively, it should refrain from downloading an update - the current approach can result in very surprising and problematic results for the user.

**Comment by robertdebock** [comment]
Hi @michaeloa, I see what you mean. I'll try to think of a fix.

**Comment by robertdebock** [comment]
Okay, the update should also restart the service, that's simple.

For the data; I think the data needs to be stored outside of the directories that host the application. (/opt/artifactory-oss-6.16.0)

I'll need to write some configuration change for that.

TL;DR: help is on the way.

**Comment by robertdebock** [comment]
I've prepared everything, CI will test the changes after which I will release a new version.

**Comment by robertdebock** [comment]
Hi, version 3.0.0 has been released, which contain fixes for the issues you identified. Thanks a lot @michaeloa for your contribution!

**Comment by michaeloa** [comment]
Awesome. Will try to get this tested on our test servers once I get a little time. Many thanks and Merry Christmas.

**Comment by weakcamel** [comment]
With full appreciation for the existing Ansible role, I don't believe it actually supports upgrade scenario at this point - it rather installs a new version in the new location.

There is much, much more to an Artifactory upgrade for a zip installation:

-  config files to be preserved/overwritten: https://www.jfrog.com/confluence/display/RTF/Upgrading+Artifactory#UpgradingArtifactory-ZIPInstallation as well as e.g. database drivers
- a "systemctl restart" likely won't as a handler: it would use a "stop" script from the **new** location rather than stopping the existing

I'm sure this role would be useful for some anyway; personally I'll be looking again at Deb/Rpm-based installation or perhaps running Artifactory as a Docker container.
An RPM/DEB based installation takes care of it

**Comment by michaeloa** [comment]
Hi. After having looked at what happened when I ran the new version of this in test, and then doing a little bit of investigation on the old server, I think @weakcamel is right, and that this still doesn't support upgrades.

In addition to the stuff that @weakcamel references above, there is also the matter of the database. In this role, you chose to use the "embedded" derby database rather than making it configurable - which is fine for my use case - but it means that you have an entire database also located with your installation. And that database is again stored with the installation - which means that it does **not** get moved if you install a new version of the system. So with the latest version, you'd have all of the files uploaded preserved, but that wouldn't matter since the database would be empty after an upgrade.

I do think this could probably be fixed relatively easily. 
1. Split the derby database out into its own directory, as done with the file storage.
2. Run the artifactory service from a "current" installation, which is symbolically linked to the latest version of the installation. This easily allows for a correct stopping of the old server, and transition to new version.
3. I don't think config files actually need to be preserved in this case, since Ansible should regenerate them correctly each time.

Can probably create a PR for this if I can find a little time to look at it this month.

**Comment by weakcamel** [comment]
That sounds like a good plan.

Step (1) might be a bit tricky as there doesn't seem to be a property in place (at least nothing documented) as to where the Derby DB has its files; I'm guessing that's just hardcoded and you might need to "cheat" with symbolic links.

**Comment by michaeloa** [comment]
Hmm - well, symbolic links would resolve the issue. Alternatively, one could also just install the zip file in the "intended" way, as described in the linked document; i.e., explicitly moving over the bin, war files, etc.
